test/libyul/StackReuseCodegen.cpp

```diff
@@ -0,0 +1,353 @@
+/*
+    This file is part of solidity.
+
+    solidity is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    solidity is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/**
+ * Unit tests for stack-reusing code generator.
+ */
+
+#include <test/Options.h>
+
+#include <libsolidity/interface/AssemblyStack.h>
+#include <libevmasm/Instruction.h>
+
+using namespace std;
+
+namespace dev
+{
+namespace yul
+{
+namespace test
+{
+
+namespace
+{
+std::string assemble(std::string const& _input)
+{
+	solidity::AssemblyStack asmStack;
+	BOOST_REQUIRE_MESSAGE(asmStack.parseAndAnalyze("", _input), "Source did not parse: " + _input);
+	return solidity::disassemble(asmStack.assemble(solidity::AssemblyStack::Machine::EVM, true).bytecode->bytecode);
+}
+}
+
+BOOST_AUTO_TEST_SUITE(StackReuseCodegen)
+
+BOOST_AUTO_TEST_CASE(smoke_test)
+{
+	string out = assemble("{}");
+	BOOST_CHECK_EQUAL(out, "");
+}
+
+BOOST_AUTO_TEST_CASE(single_var)
+{
+	string out = assemble("{ let x }");
+	BOOST_CHECK_EQUAL(out, "PUSH1 0x0 POP ");
+}
+
+BOOST_AUTO_TEST_CASE(single_var_assigned)
+{
+	string out = assemble("{ let x := 1 }");
+	BOOST_CHECK_EQUAL(out, "PUSH1 0x1 POP ");
+}
+
+BOOST_AUTO_TEST_CASE(single_var_assigned_plus_code)
+{
+	string out = assemble("{ let x := 1 mstore(3, 4) }");
+	BOOST_CHECK_EQUAL(out, "PUSH1 0x1 POP PUSH1 0x4 PUSH1 0x3 MSTORE ");
+}
+
+BOOST_AUTO_TEST_CASE(single_var_assigned_plus_code_and_reused)
+{
+	string out = assemble("{ let x := 1 mstore(3, 4) pop(mload(x)) }");
+	BOOST_CHECK_EQUAL(out, "PUSH1 0x1 PUSH1 0x4 PUSH1 0x3 MSTORE DUP1 MLOAD POP POP ");
+}
+
+BOOST_AUTO_TEST_CASE(multi_reuse_single_slot)
+{
+	string out = assemble("{ let x := 1 x := 6 let y := 2 y := 4 }");
+	BOOST_CHECK_EQUAL(out, "PUSH1 0x1 PUSH1 0x6 SWAP1 POP POP PUSH1 0x2 PUSH1 0x4 SWAP1 POP POP ");
```

Comment:
Why the `SWAP`s?

Reply:
Is it for the order of `POP`s?

Reply:
`:=` always translates into `swap + pop`. The main improvement this PR does is to reuse stack slots of other variables, but the mapping `variable -> stack slot` is still constant across the lifetime of the variable. Another improvement would be to make this mapping non-constant.

Reply:
Please also keep in mind removing redundant push/pop-combinations is the job of the yul-based optimizer - it will remove redundant assignments to the same variable.

Reply:
Also, using SSA this will probably be rewritten to `let x1 := 1 let x2 := 6 let y1 := 2 let y2 := 4` and thus it will be `push1 1 pop push1 6 pop push1 2 pop push1 4 pop`

Reply:
Right